codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/RegisterServiceMetadataMiddleware.java

```diff
@@ -41,39 +43,47 @@ public void writeAdditionalFiles(
             OperationShape operation = model.expectShape(operationId, OperationShape.class);
             goDelegator.useShapeWriter(operation, writer -> {
                 writer.openBlock("func $L(region string) $P {", "}",
                         middlewareName, serviceMetadataProvider, () -> {
-                    StringBuilder builder = new StringBuilder();
-                    builder.append(" return &$T{\n");
-                    builder.append("Region: region,\n");
-                    builder.append("ServiceID: ServiceID,\n");
+                            StringBuilder builder = new StringBuilder();
+                            builder.append(" return &$T{\n");
+                            builder.append("Region: region,\n");
+                            builder.append("ServiceID: ServiceID,\n");
 
-                    Map<ShapeId, Trait> authSchemes = serviceIndex.getEffectiveAuthSchemes(service, operation);
-                    if (authSchemes.containsKey(SigV4Trait.ID)) {
-                        SigV4Trait trait = (SigV4Trait) authSchemes.get(SigV4Trait.ID);
-                        builder.append(String.format("SigningName: \"%s\",\n", trait.getName()));
-                    }
-                    builder.append(String.format("OperationName: \"%s\",\n", operation.getId().getName(service)));
-                    builder.append("}");
+                            Map<ShapeId, Trait> authSchemes = serviceIndex.getEffectiveAuthSchemes(service, operation);
+                            if (authSchemes.containsKey(SigV4Trait.ID)) {
+                                SigV4Trait trait = (SigV4Trait) authSchemes.get(SigV4Trait.ID);
+                                builder.append(String.format("SigningName: \"%s\",\n", trait.getName()));
+                            }
+                            builder.append(String.format("OperationName: \"%s\",\n",
+                                    operation.getId().getName(service)));
+                            builder.append("}");
 
-                    writer.write(builder.toString(), serviceMetadataProvider);
-                });
+                            writer.write(builder.toString(), serviceMetadataProvider);
+                        });
             });
         }
     }
 
-        @Override
+    @Override
     public void processFinalizedModel(GoSettings settings, Model model) {
         ServiceShape service = settings.getService(model);
-        for (ShapeId operationId : service.getAllOperations()) {
-            String middlewareName = getServiceMetadataMiddlewareName(service, operationId);
-            OperationShape operation = model.expectShape(operationId, OperationShape.class);
+        TopDownIndex index = TopDownIndex.of(model);
+
+        for (ToShapeId operationId : index.getContainedOperations(service)) {
+            String middlewareName = getServiceMetadataMiddlewareName(service, operationId.toShapeId());
```

Comment:
Nit: can rename `operationId` -> `operation` - since we use ` operationId.toShapeId()`